### PR TITLE
Oauth2 for mail server

### PR DESCRIPTION
**Pull request by dashevchenko**
## Pull Request description

Added new type of authentication for mail server - oauth2. 
Added mail settings template: Google, Sendgrid, Office 365, Custom.

UI:
Screenshot from 2023-04-25 16-50-04
Screenshot from 2023-04-25 16-50-13

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: dao/src/main/java/org/thingsboard/server/dao/nosql/CassandraAbstractDao.java [hunk 1]
```diff
@@ -94,7 +94,7 @@ private TbResultSetFuture executeAsync(TenantId tenantId, Statement statement, C
             log.debug("Execute cassandra async statement {}", statementToString(statement));
         }
         if (statement.getConsistencyLevel() == null) {
-            statement.setConsistencyLevel(level);
+            statement = statement.setConsistencyLevel(level);
```
Comment (YevhenBondarenko) [inline dao/src/main/java/org/thingsboard/server/dao/nosql/CassandraAbstractDao.java:97]: how its related to the oauth2 feature?

**Review by YevhenBondarenko — APPROVED** [review]

**Review by vvlladd28 — APPROVED** [review]

File: ui-ngx/src/app/core/http/admin.service.ts [hunk 2]
```diff
@@ -131,4 +131,16 @@ export class AdminService {
...
+
+  public getLoginProcessingUrl(config?: RequestConfig): Observable<string> {
+    return this.http.get<string>(`/api/admin/mail/oauth2/loginProcessingUrl`, defaultHttpOptionsFromConfig(config));
+  }
+
+  public generateAccessToken(config?: RequestConfig): Observable<string> {
+    return this.http.get<string>(`/api/admin/mail/oauth2/authorize`, defaultHttpOptionsFromConfig(config));
+  }
+
+  public getMailConfigTemplate(config?: RequestConfig): Observable<Array<any>> {
```
Comment (vvlladd28) [inline ui-ngx/src/app/core/http/admin.service.ts:143]: Why does the function return type any?

File: ui-ngx/src/assets/locale/locale.constant-en_US.json [hunk 3]
```diff
@@ -207,7 +208,7 @@
           "login-button-label": "Provider label",
           "login-button-label-placeholder": "Login with $(Provider label)",
           "login-button-label-required": "Label is required.",
-          "login-provider": "Login provider",
+          "smtp-provider": "SMTP provider",
```
Comment (vvlladd28) [inline ui-ngx/src/assets/locale/locale.constant-en_US.json:211]: You can break ui component oauth2-settings

File: ui-ngx/src/app/shared/models/oauth2.models.ts [hunk 4]
```diff
@@ -139,3 +139,18 @@ export interface OAuth2ClientInfo {
...
+  CUSTOM = 'CUSTOM'
+}
+export const mailServerOauth2ProvidersTranslations = new Map<MailServerOauth2Provider, string>(
+  [
+    [MailServerOauth2Provider.GOOGLE, 'admin.oauth2.google-provider'],
+    [MailServerOauth2Provider.OFFICE_365, 'admin.oauth2.microsoft-provider'],
+    [MailServerOauth2Provider.SENDGRID, 'admin.oauth2.sendgrid-provider'],
+    [MailServerOauth2Provider.CUSTOM, 'admin.oauth2.custom-provider']
+  ]
+);
```
Comment (vvlladd28) [inline ui-ngx/src/app/shared/models/oauth2.models.ts:156]: This setting must be with getMailConfigTemplate

File: ui-ngx/src/app/modules/home/pages/admin/mail-server.component.ts [hunk 5]
```diff
@@ -77,70 +178,176 @@ export class MailServerComponent extends PageComponent implements OnInit, OnDest
...
   sendTestMail(): void {
     this.adminSettings.jsonValue = {...this.adminSettings.jsonValue, ...this.mailSettingsFormValue};
     this.adminService.sendTestMail(this.adminSettings).subscribe(
-      () => {
-        this.store.dispatch(new ActionNotificationShow({ message: this.translate.instant('admin.test-mail-sent'),
-          type: 'success' }));
-      }
+      () => this.store.dispatch(new ActionNotificationShow({ message: this.translate.instant('admin.test-mail-sent'),
+          type: 'success' })),
+      error => this.store.dispatch(new ActionNotificationShow({message: error.error.message, type: 'error'}))
```
Comment (vvlladd28) [inline ui-ngx/src/app/modules/home/pages/admin/mail-server.component.ts:350]: Need to rewrite in new style subscription

File: ui-ngx/src/app/modules/home/pages/admin/mail-server.component.ts [hunk 6]
```diff
@@ -151,16 +358,56 @@ export class MailServerComponent extends PageComponent implements OnInit, OnDest
...
+    const domainInfo = this.domainForm.value;
+    if (domainInfo.name !== '') {
+      const protocol = isDefined(schema) ? schema.toLowerCase() : domainInfo.scheme.toLowerCase();
+      const domainName = isDefined(name) ? name : domainInfo.name;
+      return `${protocol}://${domainName}${this.loginProcessingUrl}`;
+    }
+    return '';
+  }
+
+  private parseUrl(value: string) {
```
Comment (vvlladd28) [inline ui-ngx/src/app/modules/home/pages/admin/mail-server.component.ts:383]: What type does the function return?

File: ui-ngx/src/app/modules/home/pages/admin/mail-server.component.ts [hunk 7]
```diff
@@ -151,16 +358,56 @@ export class MailServerComponent extends PageComponent implements OnInit, OnDest
...
+    if (value) {
+      const url = new URL(value);
+      this.domainForm.get('scheme').patchValue(
+        url.protocol.startsWith('https') ? DomainSchema.HTTPS : DomainSchema.HTTP, {emitEvent: false}
+      );
+      this.domainForm.get('name').patchValue(url.host, {emitEvent: false});
+    }
+  }
+
+  get accessTokenButtonName() {
```
Comment (vvlladd28) [inline ui-ngx/src/app/modules/home/pages/admin/mail-server.component.ts:393]: What type does the function return?

File: ui-ngx/src/app/modules/home/pages/admin/mail-server.component.ts [hunk 8]
```diff
@@ -151,16 +358,56 @@ export class MailServerComponent extends PageComponent implements OnInit, OnDest
...
+    return this.translate.instant(
+      this.adminSettings.jsonValue.tokenGenerated ? 'admin.oauth2.update-access-token' : 'admin.oauth2.generate-access-token'
+    );
+  }
+
+  get accessTokenStatus() {
```
Comment (vvlladd28) [inline ui-ngx/src/app/modules/home/pages/admin/mail-server.component.ts:399]: What type does the function return?

**Review by vvlladd28 — CHANGES_REQUESTED** [review]

**Comment by ashvayka** [comment]
@dashevchenko , unfortunately, we need to postpone this feature to 3.5.1 since it is not yet ready. Please resolve conflicts and prepare this for merge to 3.5.1

**Review by vvlladd28 — APPROVED** [review]

**Pull request closed without merging**